### PR TITLE
cache and constraint preview dimensions

### PR DESCRIPTION
**Pull request by luukvbaal**
Sorry for the large amount of changes in a single commit, I'll do a review and comment.

### CODE REVIEW

File: plugins/preview-tui-ext [hunk 1]
```diff
@@ -76,6 +76,9 @@ USE_PISTOL="${USE_PISTOL:-0}"
 PAGER="${PAGER:-less -P?n -R}"
 ARCHIVES="$(echo "$NNN_ARCHIVE" | sed 's/.*(\(.*\)).*/\1/;s/|/ /g')"
 TMPDIR="${TMPDIR:-/tmp}"
+NNN_PREVIEWDIR="${NNN_PREVIEWDIR:-$XDG_CACHE_HOME}/nnn"
```
Comment (luukvbaal) [inline plugins/preview-tui-ext:79]: Allow choosing directory for generated image previews, `XDG_CACHE_HOME` default?

File: plugins/preview-tui-ext [hunk 2]
```diff
@@ -76,6 +76,9 @@ USE_PISTOL="${USE_PISTOL:-0}"
...
+NNN_PREVIEWWIDTH="${NNN_PREVIEWWIDTH:-1920}"
+NNN_PREVIEWHEIGHT="${NNN_PREVIEWHEIGHT:-1080}"
```
Comment (luukvbaal) [inline plugins/preview-tui-ext:81]: Use maximum dimensions for preview generation

File: plugins/preview-tui-ext [hunk 3]
```diff
@@ -168,19 +171,21 @@ preview_file () {
...
-        elif [ "${mimetype%%/*}" = "image" ]; then
-            image_preview "$cols" "$lines" "$1"
+        if [ "${mimetype%%/*}" = "image" ]; then
+            if [ "${mimetype#*/}" = "gif" ] && exists convert; then
+                generate_preview "$cols" "$lines" "$1" "gif"
+            elif [ ! "${mimetype#*/}" = "jpeg" ]; then
+                generate_preview "$cols" "$lines" "$1" "image"
+            else
+                image_preview "$cols" "$lines" "$1"
+            fi
```
Comment (luukvbaal) [inline plugins/preview-tui-ext:181]: Remove `xcf` case and `generate_preview` anything that is not `jpeg`

File: plugins/preview-tui-ext [hunk 4]
```diff
@@ -168,19 +171,21 @@ preview_file () {
...
         elif [ "${mimetype%%/*}" = "audio" ] && exists ffmpeg; then
             generate_preview "$cols" "$lines" "$1" "audio"
         elif [ "${mimetype%%/*}" = "video" ] && exists ffmpegthumbnailer; then
             generate_preview "$cols" "$lines" "$1" "video"
-        elif [ "$ext" = "pdf" ] && exists pdftoppm; then
+        elif [ "${mimetype#*/}" = "pdf" ] && exists pdftoppm; then
             generate_preview "$cols" "$lines" "$1" "pdf"
-        elif [ "$ext" = "epub" ] && exists gnome-epub-thumbnailer; then
+        elif [ "${mimetype#*epub}" != "$mimetype" ] && exists gnome-epub-thumbnailer; then
             generate_preview "$cols" "$lines" "$1" "epub"
```
Comment (luukvbaal) [inline plugins/preview-tui-ext:189]: Use `mimetype` where possible

File: plugins/preview-tui-ext [hunk 5]
```diff
@@ -226,19 +231,19 @@ preview_file () {
 }
 
 generate_preview() {
-    if [ ! -f "$TMPDIR/$3.jpg" ]; then
-        mkdir -p "$TMPDIR/${3%/*}"
+  if [ ! -f "$NNN_PREVIEWDIR/$3.jpg" ] || [ -n "$(find -L "$3" -newer "$NNN_PREVIEWDIR/$3.jpg")" ]; then
+        mkdir -p "$NNN_PREVIEWDIR/${3%/*}"
```
Comment (luukvbaal) [inline plugins/preview-tui-ext:235]: Only generate preview when it doesn't exist or the file is newer than the existing preview

File: plugins/preview-tui-ext [hunk 6]
```diff
@@ -250,20 +255,16 @@ generate_preview() {
...
+                    mv "$NNN_PREVIEWDIR/${3%/*}/$filename.jpg" "$NNN_PREVIEWDIR/$3.jpg" ;;
+            pdf) pdftoppm -jpeg -f 1 -singlefile "$3" "$NNN_PREVIEWDIR/$3" >/dev/null 2>&1 ;;
+            video) ffmpegthumbnailer -m -s0 -i "$3" -o "$NNN_PREVIEWDIR/$3.jpg" >/dev/null 2>&1 || rm "$NNN_PREVIEWDIR/$3.jpg" ;;
         esac
     fi
-    if [ -f "$TMPDIR/$3.jpg" ]; then
-        image_preview "$1" "$2" "$TMPDIR/$3.jpg"
+    if [ -f "$NNN_PREVIEWDIR/$3.jpg" ]; then
+        image_preview "$1" "$2" "$NNN_PREVIEWDIR/$3.jpg"
     else
```
Comment (luukvbaal) [inline plugins/preview-tui-ext:268]: Switch to `NNN_PREVIEWDIR` and constraint dimensions where possible

File: plugins/preview-tui-ext [hunk 7]
```diff
@@ -338,6 +335,8 @@ if [ "$PREVIEW_MODE" ]; then
 
     if [ "$TERMINAL" != "kitty" ] && exists ueberzug; then
         mkfifo "$FIFO_UEBERZUG"
+        trap 'ueberzug_refresh' WINCH
+        trap 'rm "$FIFO_UEBERZUG" "$PAGERPID" "$GIFPID" "$CURSEL"' INT HUP EXIT
         tail --follow "$FIFO_UEBERZUG" | ueberzug layer --silent --parser json &
```
Comment (luukvbaal) [inline plugins/preview-tui-ext:340]: Move to pre-existing if statement with same conditions

**Comment by jarun** [comment]
@leovilok please review.

**Comment by luukvbaal** [comment]
I think storing previews on disk and keeping them between reboots makes sense. Can result in a growing cache directory but I think it's worth it. 

Could also create a cleanup script to remove previews from the cache directory of which the files are no longer on disk(in the same location).

**Comment by jarun** [comment]
Do not leave user data in cache please. The user may not even know.

**Comment by luukvbaal** [comment]
What do you suggest then, keeping `NNN_PREVIEWDIR` default as `/tmp` and mentioning setting it to `XDG_CACHE_DIR` if you want to keep previews between reboots?

**Comment by jarun** [comment]
Yes, that would be fine. But do NOT cache anything by default.

**Review by leovilok — APPROVED** [review]
Looks good to me.

File: plugins/preview-tui-ext (hunk 4 above)
Comment (luukvbaal) [reply plugins/preview-tui-ext]: Apparently not all `pdf`'s / `epub`'s have the same `mimetype`, found some with `application/octet-stream`.

**Comment by jarun** [comment]
Thank you!

**Pull request merged**